### PR TITLE
Index workspace tags for Library; republish workspace when tags change [GAWB-2028]

### PR DESCRIPTION
**Pull request by aednichols**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Product Owner** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala [hunk 1]
```diff
@@ -107,8 +108,16 @@ trait WorkspaceApiService extends HttpService with FireCloudRequestBuilding
             }
           } ~
           path("updateAttributes") {
-            requireUserInfo() { _ =>
-              passthrough(workspacePath, HttpMethods.PATCH)
+            patch {
+              requireUserInfo() { userInfo: UserInfo =>
+                entity(as[Seq[AttributeUpdateOperation]]) { replacementAttributes => requestContext =>
+                  perRequest(requestContext,
```
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala:114]: This changed from `passthrough` to `perRequest` to accommodate the `republishIfPublished` call

File: src/main/resources/library/attribute-definitions.json [hunk 2]
```diff
@@ -313,6 +313,14 @@
...
       "indexable": false,
       "type": "integer"
+    },
+    "tag:tags": {
+      "title": "Tags",
+      "type": "array",
+      "items": { "type": "string" },
+      "aggregate": {
+        "renderHint": "tag-autocomplete"
+      }
```
Comment (ahaessly) [inline src/main/resources/library/attribute-definitions.json]: maybe use a hint that's more generic like "typeahead-multiselect"

**Review by ahaessly — APPROVED** [review]
just a suggestion

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryServiceSupport.scala [hunk 3]
```diff
@@ -76,7 +76,13 @@ trait LibraryServiceSupport extends LazyLogging {
...
       AttributeName.withDefaultNS("authorizationDomain") -> AttributeValueList(workspace.authorizationDomain.map(group => AttributeString(group.membersGroupName.value)).toSeq)
     )
-    val fields = attrfields ++ idfields
+
+    val tags: Option[Attribute] = workspace.attributes.get(AttributeName.withTagsNS())
+
+    val fields = attrfields ++ idfields ++ (tags match {
+      case Some(t) => Map(AttributeName.withTagsNS() -> t)
+      case None => Map()
+    })
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryServiceSupport.scala]: for code cleanliness: `attrfields` and `idfields` are both already `Map`s. I would make the tag code follow the same pattern by moving the match earlier:

```
    val tagfields = workspace.attributes.get(AttributeName.withTagsNS()) match {
      case Some(t) => Map(AttributeName.withTagsNS() -> t)
      case None => Map.empty[AttributeName, Attribute]
    }

    val fields = attrfields ++ idfields ++ tagfields
```

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspacePublishingSupport.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    searchDAO.deleteDocument(ws.workspaceId)
+  }
+
+  def republishIfPublished(ws: Workspace, ontologyDAO: OntologyDAO, searchDAO: SearchDAO): Unit = {
+    if (isPublished(ws)) {
+      // if already published, republish
+      // we do not need to delete before republish
+      publishDocument(ws, ontologyDAO, searchDAO)
+    }
+  }
```
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspacePublishingSupport.scala:30]: rename this to `republishDocument` to be consistent with `publishDocument` and `removeDocument`?

To me, "republish" implies "already published"

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 5]
```diff
@@ -848,6 +825,44 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService {
...
+                                                        |  }
+                                                        |]""".stripMargin))
+            ~> dummyUserIdHeaders("1234")
+            ~> sealRoute(workspaceRoutes)) ~> check {
+            status should equal(OK)
+          }
+        }
+      }
+    }
+
```
Comment (davidangb) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala]: I'd love to see some tests that actually check publishing when calling the attribute endpoints!

**Review by ahaessly — APPROVED** [review]
👍  tests look great to me

**Pull request merged**